plugins/inputs/huawei_routers_telemetry/huawei_routers_telemetry.go

```diff
@@ -0,0 +1,230 @@
+package huawei_routers_telemetry
+
+import (
+	"fmt"
+	"github.com/DamRCorba/huawei_telemetry_sensors"
+	"github.com/DamRCorba/huawei_telemetry_sensors/sensors/huawei-telemetry"
+	"github.com/golang/protobuf/proto"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"io"
+	"net"
+	"os"
+	"strings"
+	"sync"
+	"time"
+)
+
+type setReadBufferer interface {
+	SetReadBuffer(bytes int) error
+}
+
+type streamSocketListener struct {
+	net.Listener
+	*HuaweiRoutersTelemetry
+
+	sockType string
+
+	connections    map[string]net.Conn
+	connectionsMtx sync.Mutex
+}
+
+type packetSocketListener struct {
+	net.PacketConn
+	*HuaweiRoutersTelemetry
+}
+
+/*
+  Telemetry Decoder.
+
+*/
+func HuaweiTelemetryDecoder(body []byte) *metric.SeriesGrouper {
+	msg := &telemetry.Telemetry{}
+	err := proto.Unmarshal(body[12:], msg)
+	if err != nil {
+		fmt.Println("Error", err)
+		panic(err)
+	}
+	grouper := metric.NewSeriesGrouper()
+	for _, gpbkv := range msg.GetDataGpb().GetRow() {
+		dataTime := gpbkv.Timestamp
+		if dataTime == 0 {
+			dataTime = msg.MsgTimestamp
+		}
+		timestamp := time.Unix(int64(dataTime/1000), int64(dataTime%1000)*1000000)
+		sensorMsg := huawei_sensorPath.GetMessageType(msg.GetSensorPath())
+		err = proto.Unmarshal(gpbkv.Content, sensorMsg)
+		if err != nil {
+			fmt.Println("Sensor Error", err)
+		} else {
+			fields, vals := huawei_sensorPath.SearchKey(gpbkv, msg.GetSensorPath())
+			tags := make(map[string]string, len(fields)+3)
+			tags["source"] = msg.GetNodeIdStr()
+			tags["subscription"] = msg.GetSubscriptionIdStr()
+			tags["path"] = msg.GetSensorPath()
+			// Search for Tags
+			for i := 0; i < len(fields); i++ {
+				tags = huawei_sensorPath.AppendTags(fields[i], vals[i], tags, msg.GetSensorPath())
+			}
+			// Create Metrics
+			for i := 0; i < len(fields); i++ {
+				huawei_sensorPath.CreateMetrics(grouper, tags, timestamp, msg.GetSensorPath(), fields[i], vals[i])
+			}
+		}
+	}
+	return grouper
+}
+
+/*
+  Listen UDP packets and call the telemetryDecoder.
+*/
+func (psl *packetSocketListener) listen() {
+	buf := make([]byte, 64*1024) // 64kb - maximum size of IP packet
+	for {
+		n, _, err := psl.ReadFrom(buf)
+		if err != nil {
+			if !strings.HasSuffix(err.Error(), ": use of closed network connection") {
+				psl.Log.Error(err.Error())
+			}
+			break
```

Comment:
I'm not sure this is a good idea as this means the plugin will go defunct after the first read error.

Reply:
maybe i misunderstood, if the packet could not be read the loop must not continue. If i use break the plugin stop? I am using like this in a "production enviroment" and its work fine.

Reply:
What I mean is: Once you encounter a read error, you will leave the eternal `for` loop and `listen` will end. Now nothing is listening for packets anymore, so if the router side recovers or network is back or whatever, you will not receive new data anymore until you restart the plugin (and/or telegraf). That's bad if you rely on metrics coming in.

To solve this, the plugin should close the connection and try to listen again (like do an internal restart) in order to be prepared for the router-side to reconnect.